src/dbnode/storage/index/block.go

```diff
@@ -477,7 +478,7 @@ func (b *block) queryWithSpan(
 	}()
 
 	for iter.Next() {
-		if opts.LimitExceeded(size) {
+		if opts.SeriesLimitExceeded(size) || opts.DocsLimitExceeded(docsCount) {
```

Comment:
I'm not sure if this would be significant (guess it depends on `batchSize`), but `size` and `docsCount` only get increased after the batch gets full, meaning that those limit checks here are not capturing the actual resource increases as long as `len(batch) < batchSize` (see an `if` with `continue` below).
Tracking the `size` might be difficult without calling the `addQueryResults`, but doing that for `docsCount` should be trivial. Again, not really sure if this would be significant.

Reply:
Good call, but make sure to add a check that the limit hasn't been exceeded before the iteration here

Reply:
Yeah nice. So I think we just need `docsCount++` after each `batch = append(batch, iter.Current())` instead of `docsCount += len(batch)` below.